api/campaigns/class-maybe-show-campaign.php

```diff
@@ -73,27 +73,42 @@ function ( $post_data ) use ( $visit ) {
 			);
 		}
 
-		$page_referer_url = isset( $_REQUEST['ref'] ) ? $_REQUEST['ref'] : ''; // phpcs:ignore WordPress.Security.NonceVerification.Recommended, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
-
+		$referer_url                   = filter_input( INPUT_SERVER, 'HTTP_REFERER', FILTER_SANITIZE_STRING );
+		$page_referer_url              = isset( $_REQUEST['ref'] ) ? $_REQUEST['ref'] : ''; // phpcs:ignore WordPress.Security.NonceVerification.Recommended, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+		$all_segments                  = isset( $settings->all_segments ) ? $settings->all_segments : [];
+		$best_priority_segment         = $this->get_best_priority_segment( $all_segments, $client_id, $referer_url, $page_referer_url, $view_as_spec );
 		$overlay_to_maybe_display      = null;
 		$above_header_to_maybe_display = null;
 
-		foreach ( $campaigns as $campaign ) {
+		// Filter campaigns to only those that match the best priority segment, or the default segment.
+		$matching_campaigns = array_filter(
+			$campaigns,
+			function( $campaign ) use ( $best_priority_segment ) {
+				return ( $best_priority_segment === $campaign->s || empty( $campaign->s ) );
+			}
+		);
+
+		$this->debug['campaigns'] = $matching_campaigns;
+
+		// Check each matching campaign against other global factors.
+		foreach ( $matching_campaigns as $campaign ) {
 			$campaign_should_be_shown = $this->should_campaign_be_shown(
 				$client_id,
 				$campaign,
 				$settings,
 				filter_input( INPUT_SERVER, 'HTTP_REFERER', FILTER_SANITIZE_STRING ),
-				$page_referer_url,
-				$view_as_spec
+				'',
```

Comment:
Why is the `$page_referer_url` argument hard-coded to `''` here?

Reply:
This was a mistake. Fixed in 9f52331.

Also fixed in #451 here: https://github.com/Automattic/newspack-popups/pull/451/files#diff-fab5ae21610a1add85a1640bebc4871f8fa855dd8f7c88d149e2f1c4c20df540R103